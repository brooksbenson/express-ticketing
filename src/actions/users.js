import db, { auth } from '../firebase/firebase';

export const addUser = ({ key, ...user }) => ({
  type: 'ADD_USER',
  key,
  user
});

export const startAddUser = user => {
  return async dispatch => {
    const { key } = await db.ref('user_data').push(user);
    dispatch(addUser({ key, ...user }));
    return key;
  };
};

export const setUsers = users => ({
  type: 'SET_USERS',
  users
});

export const startSetUsers = () => {
  return async dispatch => {
    const snap = await db.ref('user_data').once('value');
    dispatch(setUsers(snap.val()));
  };
};
