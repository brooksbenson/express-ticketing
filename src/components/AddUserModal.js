import React from 'react';
import Modal from 'react-modal';
import SearchBar from './SearchBar';
import userSelector from '../selectors/users';

Modal.setAppElement('body');

export default props => (
  <Modal
    className="add-user-modal"
    isOpen={props.isOpen}
    contentLabel="Add user modal"
    onRequestClose={props.onRequestClose}
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    closeTimeoutMS={200}
  >
    <h3> Add User </h3>
    <button onClick={props.onRequestClose}>
      <div />
      <div />
    </button>
    <SearchBar
      displayResults={!!props.searchString}
      onPick={props.onUserPick}
      onSearchChange={e => props.onSearchChange(e.target.value)}
      results={userSelector(props.users, props.searchString)}
      searchString={props.searchString}
    />
    <button onClick={props.onAddUser}>Add User</button>
  </Modal>
);
