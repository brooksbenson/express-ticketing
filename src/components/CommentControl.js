import React from 'react';

export default props => (
  <section className={props.className}>
    <h3> Comments </h3>
    <textarea
      onChange={e => props.onCommentChange(e.target.value)}
      value={props.comment}
    />
    <button onClick={props.onCommentSave}> Save Comment </button>
    <ul>
      {props.comments.map(c => (
        <li key={c.key}>
          <div>
            <span>{c.name}</span>
            <span>{c.date}</span>
          </div>
          <p>{c.text}</p>
        </li>
      ))}
    </ul>
  </section>
);
