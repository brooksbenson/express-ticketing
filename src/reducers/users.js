export default (state = {}, action) => {
  switch (action.type) {
    case 'ADD_USER':
      return { ...state, [action.key]: action.user };
    case 'SET_USERS':
      return action.users;
    default:
      return state;
  }
};
