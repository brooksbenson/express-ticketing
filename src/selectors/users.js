export default (users, search) => {
  const regex = new RegExp(search.split('').join('.*?'), 'i');
  return users.filter(user => regex.test(user.email));
};
