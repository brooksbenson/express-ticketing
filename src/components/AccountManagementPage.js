import React from 'react';
import { connect } from 'react-redux';
import SearchList from './reuse/SearchList';
import { startAddAccount, startUpdateAccount } from '../actions/accounts';
import accountSelector from '../selectors/accounts';

const stateDefaults = {
  key: '',
  name: '',
  website: ''
};

export class AccountManagementPage extends React.Component {
  state = { ...stateDefaults };

  onAccountClick = account => {
    this.setState(() => ({ ...account }));
  };

  onCancel = () => {
    this.setState(() => ({ ...stateDefaults }));
  };

  onNameChange = e => {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };

  onWebsiteChange = e => {
    const website = e.target.value;
    this.setState(() => ({ website }));
  };

  onSave = e => {
    e.preventDefault();
    if (this.state.key) {
      this.props.startUpdateAccount({ ...this.state }).then(() => {
        this.setState(() => ({ ...stateDefaults }));
      });
    } else {
      const { key, ...accountData } = this.state;
      this.props.startAddAccount(accountData).then(() => {
        this.setState(() => ({ ...stateDefaults }));
      });
    }
  };

  onSearch = e => {
    const search = e.target.value;
    this.setState(() => ({ search }));
  };

  render() {
    const { key, name, website } = this.state;
    const { accounts } = this.props;
    return (
      <section className="content-container">
        <div className="manage-accounts content-innards">
          <h2 className="heading"> Manage Accounts </h2>
          <div className="manage-accounts__block">
            <h3 className="heading heading--secondary">
              {key ? 'Update' : 'New'} Account
            </h3>
            <form className="manage-accounts__form" onSubmit={this.onSave}>
              <div className="manage-accounts__form-row">
                <span> Name </span>
                <input
                  className="input"
                  onChange={this.onNameChange}
                  type="text"
                  value={name}
                />
              </div>
              <div className="manage-accounts__form-row">
                <span> Website </span>
                <input
                  className="input"
                  onChange={this.onWebsiteChange}
                  type="text"
                  value={website}
                />
              </div>
              <button className="btn btn--secondary">Save</button>
              {key && (
                <button className="btn btn-tertiary" onClick={this.onCancel}>
                  Cancel
                </button>
              )}
            </form>
          </div>
          <div className="manage-accounts__block">
            <h3 className="heading heading--secondary"> Accounts </h3>
            <SearchList
              className="manage-accounts__list"
              list={accounts}
              onClick={this.onAccountClick}
              selector={accountSelector}
            />
          </div>
        </div>
      </section>
    );
  }
}

const mapStateToProps = ({ accounts }) => ({ accounts });
const mapDispatchToProps = dispatch => ({
  startAddAccount: async account => {
    await dispatch(startAddAccount(account));
  },
  startUpdateAccount: async update => {
    await dispatch(startUpdateAccount(update));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(
  AccountManagementPage
);
