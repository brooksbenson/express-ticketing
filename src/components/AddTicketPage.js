import React from 'react';
import { connect } from 'react-redux';
import SearchBar from './SearchBar';
import ContactController from './ContactController';
import { startAddTicket } from '../actions/tickets';
import {
  accountPick,
  accountSearchChange,
  contactCtrlDataChange,
  contactPick,
  contactSearchChange,
  descriptionChange,
  startNewContact,
  startUpdateContact,
  titleChange,
  toggleContactCtrl,
  urgencyChange
} from '../actions/new-ticket';

export const AddTicketPage = props => {
  const onUrgencyChange = e => {
    const change = e.target.value;
    if (/^(low|medium|high)$/i.test(change)) {
      props.urgencyChange(change);
    }
  };

  const onTitleChange = e => {
    const change = e.target.value;
    if (change.length <= 40) {
      props.titleChange(change);
    }
  };

  const submit = e => {
    e.preventDefault();
    props.startAddTicket({
      accountName: props.account.name,
      contactName: props.contact.name,
      contactEmail: props.contact.email,
      contactNumber: props.contact.number,
      assignments: [{ uid: props.user.uid, completed: false }],
      comments: [{ uid: props.user.uid, text: props.description }],
      date: Date.now(),
      status: 'Open',
      title: props.title,
      urgency: props.urgency
    });
  };

  return (
    <div className="ticket-form">
      <h2 className="heading"> Create Ticket </h2>
      <section className="ticket-form__block">
        <h3 className="heading-secondary"> Account </h3>
        <SearchBar
          className="ticket-form__search-bar"
          disabled={false}
          displayKey="name"
          onPick={props.accountPick}
          onSearchChange={props.accountSearchChange}
          placeholder="Search accounts..."
          searchString={props.accountSearchString}
          results={props.accountSearchResults}
          uniqueKey={'id'}
          values={props.accounts}
        />
        <div className="ticket-form__block-row">
          <SearchBar
            className="ticket-form__search-bar"
            disabled={props.account == null}
            displayKey="name"
            onPick={props.contactPick}
            onSearchChange={props.contactSearchChange}
            placeholder="Search contacts..."
            searchString={props.contactSearchString}
            results={props.contactSearchResults}
            uniqueKey={'email'}
            values={props.account ? props.account.contacts : []}
          />
          <ContactController {...props} />
        </div>
      </section>
      <section className="ticket-form__block">
        <h3 className="heading-secondary">Triage</h3>
        <div className="ticket-form__block-row">
          <input
            className="input"
            name="title"
            onChange={onTitleChange}
            placeholder="Title"
            type="text"
          />
          <select
            className="select"
            onChange={onUrgencyChange}
            value={props.urgency || 'urgency'}
          >
            <option value="urgency" disabled>
              Urgency
            </option>
            <option value="low">Low</option>
            <option value="medium">Medium</option>
            <option value="high">High</option>
          </select>
        </div>
        <textarea
          className="textarea ticket-form__textarea"
          name="description"
          onChange={e => props.descriptionChange(e.target.value)}
          placeholder="Description (optional)"
          value={props.description}
        />
      </section>
      <button className="btn btn--primary" onClick={submit}>
        Submit Ticket
      </button>
    </div>
  );
};

const mapStateToProps = ({ newTicket, user }) => ({ ...newTicket, user });
const mapDispatchToProps = dispatch => ({
  accountPick: pick => dispatch(accountPick(pick)),
  accountSearchChange: change => dispatch(accountSearchChange(change)),
  addTicket: ticket => dispatch(addTicket(ticket)),
  contactCtrlDataChange: change => dispatch(contactCtrlDataChange(change)),
  contactPick: pick => dispatch(contactPick(pick)),
  contactSearchChange: change => dispatch(contactSearchChange(change)),
  descriptionChange: change => dispatch(descriptionChange(change)),
  startNewContact: contact => dispatch(startNewContact(contact)),
  startUpdateContact: contact => dispatch(startUpdateContact(contact)),
  titleChange: change => dispatch(titleChange(change)),
  toggleContactCtrl: () => dispatch(toggleContactCtrl()),
  urgencyChange: change => dispatch(urgencyChange(change))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTicketPage);
