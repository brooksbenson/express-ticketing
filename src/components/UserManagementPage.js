import React from 'react';
import { connect } from 'react-redux';
import { auth } from '../firebase/firebase';
import { startAddUser } from '../actions/users';
import userSelector from '../selectors/users';

export class UserManagementPage extends React.Component {
  state = {
    email: '',
    password: '',
    name: '',
    admin: false,
    search: ''
  };

  onEmailChange = e => {
    const email = e.target.value;
    this.setState(() => ({ email }));
  };

  onPasswordChange = e => {
    const password = e.target.value;
    this.setState(() => ({ password }));
  };

  onNameChange = e => {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };

  onAdminCheckboxClick = () => {
    this.setState(({ admin }) => ({ admin: !admin }));
  };

  onSave = e => {
    e.preventDefault();
    const { admin, email, password, name } = this.state;
    auth.createUserWithEmailAndPassword(email, password).then(() => {
      this.props.startAddUser({ admin, email, name }).then(() => {
        this.setState(() => ({
          admin: false,
          name: '',
          email: '',
          password: ''
        }));
      });
    });
  };

  onSearch = e => {
    const search = e.target.value;
    this.setState(() => ({ search }));
  };

  render() {
    const { email, password, admin, search, name } = this.state;
    return (
      <section className="content-container">
        <div className="manage-users content-innards">
          <h2 className="heading"> Manage Users </h2>
          <div className="manage-users__block">
            <h3 className="heading heading--secondary">New User</h3>
            <form className="manage-users__form" onSubmit={this.onSave}>
              <div className="manage-users__form-row">
                <span> Name </span>
                <input
                  className="input"
                  onChange={this.onNameChange}
                  type="text"
                  value={name}
                />
              </div>
              <div className="manage-users__form-row">
                <span> Email </span>
                <input
                  className="input"
                  onChange={this.onEmailChange}
                  type="text"
                  value={email}
                />
              </div>
              <div className="manage-users__form-row">
                <span> Password </span>
                <input
                  className="input"
                  onChange={this.onPasswordChange}
                  type="text"
                  value={password}
                />
              </div>

              <div className="manage-users__admin-checkbox">
                <label htmlFor="admin"> Admin </label>
                <input
                  name="admin"
                  onClick={this.onAdminCheckboxClick}
                  type="checkbox"
                  value={admin}
                />
              </div>
              <button className="btn btn--secondary">Save</button>
            </form>
          </div>
          <div className="manage-users__block">
            <h3 className="heading heading--secondary"> Users </h3>
            <input
              className="search"
              onChange={this.onSearch}
              placeholder="Search..."
              type="text"
              value={search}
            />
            <ul className="manage-users__list">
              {userSelector(this.props.users, search).map(user => (
                <li key={user.key}> {user.email} </li>
              ))}
            </ul>
          </div>
        </div>
      </section>
    );
  }
}

const mapStateToProps = ({ users }) => ({ users });
const mapDispatchToProps = dispatch => ({
  startAddUser: async user => {
    await dispatch(startAddUser(user));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(UserManagementPage);
