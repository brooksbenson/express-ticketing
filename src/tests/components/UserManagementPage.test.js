import React from 'react';
import { shallow } from 'enzyme';
import { UserManagementPage } from '../../components/UserManagementPage';
import { usersObj } from '../fixtures/users';

const users = Object.keys(usersObj).map(key => ({ key, ...usersObj[key] }));

let wrapper;
let startAddUserArg;
beforeEach(() => {
  wrapper = shallow(<UserManagementPage users={users} />);
});

test('UserManagementPage should match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});

test('UserManagementPage should setState on email change', () => {
  const value = 'Brooks Benson';
  wrapper
    .find('input')
    .at(0)
    .simulate('change', { target: { value } });
  expect(wrapper.state('name')).toBe(value);
});

test('UserManagementPage should setState on email change', () => {
  const value = 'brooks@mail.com';
  wrapper
    .find('input')
    .at(1)
    .simulate('change', { target: { value } });
  expect(wrapper.state('email')).toBe(value);
});

test('UserManagementPage should setState on password change', () => {
  const value = 'password1';
  wrapper
    .find('input')
    .at(2)
    .simulate('change', { target: { value } });
  expect(wrapper.state('password')).toBe(value);
});

test('should setState on admin checkbox click', () => {
  const checkboxInput = wrapper.find('input[type="checkbox"]');
  checkboxInput.simulate('click');
  expect(wrapper.state('admin')).toBeTruthy();
  checkboxInput.simulate('click');
  expect(wrapper.state('admin')).toBeFalsy();
});
