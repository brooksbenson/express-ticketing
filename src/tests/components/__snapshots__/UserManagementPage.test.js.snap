// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserManagementPage should match snapshot 1`] = `
<section
  className="content-container"
>
  <div
    className="manage-users content-innards"
  >
    <h2
      className="heading"
    >
       Manage Users 
    </h2>
    <div
      className="manage-users__block"
    >
      <h3
        className="heading heading--secondary"
      >
        New User
      </h3>
      <form
        className="manage-users__form"
        onSubmit={[Function]}
      >
        <div
          className="manage-users__form-row"
        >
          <span>
             Name 
          </span>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="manage-users__form-row"
        >
          <span>
             Email 
          </span>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="manage-users__form-row"
        >
          <span>
             Password 
          </span>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="manage-users__admin-checkbox"
        >
          <label
            htmlFor="admin"
          >
             Admin 
          </label>
          <input
            name="admin"
            onClick={[Function]}
            type="checkbox"
            value={false}
          />
        </div>
        <button
          className="btn btn--secondary"
        >
          Save
        </button>
      </form>
    </div>
    <div
      className="manage-users__block"
    >
      <h3
        className="heading heading--secondary"
      >
         Users 
      </h3>
      <input
        className="search"
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value=""
      />
      <ul
        className="manage-users__list"
      >
        <li
          key="kdlejf"
        >
           
          max@mail.com
           
        </li>
        <li
          key="-keiajf"
        >
           
          brooks@mail.com
           
        </li>
        <li
          key="-wqqadv"
        >
           
          nick@mail.com
           
        </li>
      </ul>
    </div>
  </div>
</section>
`;
